File: absoluteOrientation.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%  Code for Absolute Orientation of dataset B to dataset A, both with n points each, where for each i in {1,2,...,n}, points A_i and B_i correspond

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% loading datasets : Here, I just create two random matrices of identical dimensions
A = rand(100,30);
B = rand(100,30):

% centering 
Aprime=A-mean(A);
Bprime=B-mean(B);

% rotation
N=zeros(size(A,2),size(A,2));

for i = 1:size(A,1)
	N=N + Bprime(i,:)'*Aprime(i,:);
end

[u,s,v] = svd(N);
R = v*(u');

B_rot = Bprime*(R');

% finding the scale (optional)
s = 1.0; s1 = 0.0; s2 = 0.0;
for i=1:size(A,1)
	s1 = s1 + dot(Aprime(i,:),B_rot(i,:));
	s2 = s2 + dot(Bprime(i,:),Bprime(i,:));  %dot product of B_i
end

s=s1/s2;

% getting the rotated, translated and scaled matrix
newB = s*B_rot + mean(A)

File: absoluteOrientation.py
##########################################################################################################################################

# Python code for Absolute Orientation of dataset B to dataset A, both with n points each, where for each i in {1,2,...,n}, points A_i and B_i correspond

##########################################################################################################################################

import numpy as np


# loading datasets : Dummy matrices initiated here. You can load accordingly.
A = np.random.rand(100,30)
B = np.random.rand(100,30)

# centering 
Aprime = A - np.mean(A,0)
Bprime = B - np.mean(B,0)

# rotation
N = np.zeros((np.shape(A)[1],np.shape(A)[1]));
for i in range(m):
	N = N + np.outer(Bprime[i],Aprime[i])

U,S,V = np.linalg.svd(N1)
R = np.matmul(U,V)
newBprime = np.matmul(Bprime, R1) 

# scaling
s1=0.0
s2=0.0

for i in range(0,m):
	s1=s1+np.dot(Aprime[i],newBprime[i])
	s4=s4+np.dot(Bprime[i], Bprime[i])

s = s1/s2;

# output matrix : B oriented onto A
newB = s*newBprime + np.mean(A,0)
